File: src/uFire_ISE.cpp
// Copyright (c) 2018-2020 Justin Decker

//
// MIT License
//
// Permission is hereby granted, free of charge, to any person obtaining
// a copy of this software and associated documentation files (the
// "Software"), to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to
// permit persons to whom the Software is furnished to do so, subject to
// the following conditions:
//
// The above copyright notice and this permission notice shall be
// included in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

#include <math.h>
#include "uFire_ISE.h"

bool uFire_ISE::begin(uint8_t address, TwoWire &wirePort)
{
  _address = address;
  _i2cPort = &wirePort;

  return connected();
}

float uFire_ISE::measuremV()
{
  _send_command(ISE_MEASURE_MV);
  if(_blocking) delay(ISE_MV_MEASURE_TIME);
  _updateRegisters();

  return mV;
}

float uFire_ISE::measureTemp()
{
  _send_command(ISE_MEASURE_TEMP);
  if(_blocking) delay(ISE_TEMP_MEASURE_TIME);
  _updateRegisters();

  return tempC;

}

void uFire_ISE::setTemp(float temp_C)
{
  _write_register(ISE_TEMP_REGISTER, temp_C);
  tempC = temp_C;
  tempF = ((tempC * 9) / 5) + 32;
}

float uFire_ISE::calibrateSingle(float solutionmV)
{
  _write_register(ISE_SOLUTION_REGISTER, solutionmV);
  _send_command(ISE_CALIBRATE_SINGLE);
  if(_blocking) delay(ISE_MV_MEASURE_TIME);

  return getCalibrateOffset();
}

float uFire_ISE::calibrateProbeLow(float solutionmV)
{
  _write_register(ISE_SOLUTION_REGISTER, solutionmV);
  _send_command(ISE_CALIBRATE_LOW);
  if(_blocking) delay(ISE_MV_MEASURE_TIME);

  return getCalibrateLowReading();
}

float uFire_ISE::calibrateProbeHigh(float solutionmV)
{
  _write_register(ISE_SOLUTION_REGISTER, solutionmV);
  _send_command(ISE_CALIBRATE_HIGH);
  if(_blocking) delay(ISE_MV_MEASURE_TIME);

  return getCalibrateHighReading();
}

void uFire_ISE::setDualPointCalibration(float refLow,
                                        float refHigh,
                                        float readLow,
                                        float readHigh)
{
  _write_register(ISE_CALIBRATE_REFLOW_REGISTER,   refLow);
  _write_register(ISE_CALIBRATE_REFHIGH_REGISTER,  refHigh);
  _write_register(ISE_CALIBRATE_READLOW_REGISTER,  readLow);
  _write_register(ISE_CALIBRATE_READHIGH_REGISTER, readHigh);
}

float uFire_ISE::getCalibrateOffset()
{
  return _read_register(ISE_CALIBRATE_SINGLE_REGISTER);
}

float uFire_ISE::getCalibrateHighReference()
{
  return _read_register(ISE_CALIBRATE_REFHIGH_REGISTER);
}

float uFire_ISE::getCalibrateHighReading()
{
  return _read_register(ISE_CALIBRATE_READHIGH_REGISTER);
}

float uFire_ISE::getCalibrateLowReference()
{
  return _read_register(ISE_CALIBRATE_REFLOW_REGISTER);
}

float uFire_ISE::getCalibrateLowReading()
{
  return _read_register(ISE_CALIBRATE_READLOW_REGISTER);
}

void uFire_ISE::useTemperatureCompensation(bool b)
{
  uint8_t retval;
  uint8_t config = _read_byte(ISE_CONFIG_REGISTER);

  if (b)
  {
    retval = bitSet(config, ISE_TEMP_COMPENSATION_CONFIG_BIT);
  }
  else
  {
    retval = bitClear(config, ISE_TEMP_COMPENSATION_CONFIG_BIT);
  }
  _write_byte(ISE_CONFIG_REGISTER, retval);
}

uint8_t uFire_ISE::getVersion()
{
  return _read_byte(ISE_VERSION_REGISTER);
}

uint8_t uFire_ISE::getFirmware()
{
  return _read_byte(ISE_FW_VERSION_REGISTER);
}

void uFire_ISE::reset()
{
  _write_register(ISE_CALIBRATE_SINGLE_REGISTER, NAN);
  delay(10);
  _write_register(ISE_CALIBRATE_REFHIGH_REGISTER, NAN);
  delay(10);
  _write_register(ISE_CALIBRATE_REFLOW_REGISTER, NAN);
  delay(10);
  _write_register(ISE_CALIBRATE_READHIGH_REGISTER, NAN);
  delay(10);
  _write_register(ISE_CALIBRATE_READLOW_REGISTER, NAN);
  delay(10);
  useTemperatureCompensation(false);
}

void uFire_ISE::setI2CAddress(uint8_t i2cAddress)
{
  _write_register(ISE_SOLUTION_REGISTER, i2cAddress);
  _send_command(ISE_I2C);
  _address = i2cAddress;
}

float uFire_ISE::readEEPROM(uint8_t address)
{
  _write_register(ISE_SOLUTION_REGISTER, address);
  _send_command(ISE_MEMORY_READ);
  return _read_register(ISE_BUFFER_REGISTER);
}

void uFire_ISE::writeEEPROM(uint8_t address, float value)
{
  _write_register(ISE_SOLUTION_REGISTER, address);
  _write_register(ISE_BUFFER_REGISTER,   value);
  _send_command(ISE_MEMORY_WRITE);
}

bool uFire_ISE::connected()
{
  uint8_t retval = _read_byte(ISE_VERSION_REGISTER);

  if (retval != 0xFF) {
    return true;
  }
  else {
    return false;
  }
}

void uFire_ISE::setBlocking(bool b)
{
   _blocking = b;
}

bool uFire_ISE::getBlocking()
{
    return _blocking;
}

void uFire_ISE::readData()
{
  _updateRegisters();
  getCalibrateHighReading();
  getCalibrateHighReference();
  getCalibrateLowReading();
  getCalibrateLowReference();
  getCalibrateOffset();
}

void uFire_ISE::_updateRegisters()
{
  mV = _read_register(ISE_MV_REGISTER);
  if (isinf(mV)) {
    mV = -1;
  }
  if (isnan(mV)) {
    mV = -1;
  }

  tempC = _read_register(ISE_TEMP_REGISTER);
  if (tempC == -127.0)
  {
    tempF = -127;
  }
  else
  {
    tempF = ((tempC * 9) / 5) + 32;
  }

}

void uFire_ISE::_change_register(uint8_t r)
{
  Wire.beginTransmission(_address);
  Wire.write(r);
  Wire.endTransmission();
  delay(10);
}

void uFire_ISE::_send_command(uint8_t command)
{
  Wire.beginTransmission(_address);
  Wire.write(ISE_TASK_REGISTER);
  Wire.write(command);
  Wire.endTransmission();
  delay(10);
}

void uFire_ISE::_write_register(uint8_t reg, float f)
{
  uint8_t b[5];
  float   f_val = f;

  b[0] = reg;
  b[1] = *((uint8_t *)&f_val);
  b[2] = *((uint8_t *)&f_val + 1);
  b[3] = *((uint8_t *)&f_val + 2);
  b[4] = *((uint8_t *)&f_val + 3);
  Wire.beginTransmission(_address);
  Wire.write(b, 5);
  Wire.endTransmission();
  delay(10);
}

float uFire_ISE::_read_register(uint8_t reg)
{
  float retval;

  _change_register(reg);

  Wire.requestFrom(_address, (uint8_t)1);
  *((uint8_t *)&retval) = Wire.read();
  Wire.requestFrom(_address, (uint8_t)1);
  *((uint8_t *)&retval + 1) = Wire.read();
  Wire.requestFrom(_address, (uint8_t)1);
  *((uint8_t *)&retval + 2) = Wire.read();
  Wire.requestFrom(_address, (uint8_t)1);
  *((uint8_t *)&retval + 3) = Wire.read();
  delay(10);
  return retval;
}

void uFire_ISE::_write_byte(uint8_t reg, uint8_t val)
{
  uint8_t b[5];

  b[0] = reg;
  b[1] = val;
  Wire.beginTransmission(_address);
  Wire.write(b, 2);
  Wire.endTransmission();
  delay(10);
}

uint8_t uFire_ISE::_read_byte(uint8_t reg)
{
  uint8_t retval;

  _change_register(reg);
  Wire.requestFrom(_address, (uint8_t)1);
  retval = Wire.read();
  delay(10);
  return retval;
}

File: python/RaspberryPi/uFire_ISE.py
import math
import struct
import time
import smbus  # pylint: disable=E0401

global i2c

ISE_PROBE = 0x3F

ISE_MEASURE_MV = 80
ISE_MEASURE_TEMP = 40
ISE_CALIBRATE_SINGLE = 20
ISE_CALIBRATE_LOW = 10
ISE_CALIBRATE_HIGH = 8
ISE_MEMORY_WRITE = 4
ISE_MEMORY_READ = 2
ISE_I2C = 1

ISE_VERSION_REGISTER = 0             # hardware version */
ISE_MV_REGISTER = 1                  # mV */
ISE_TEMP_REGISTER = 5                # temperature in C */
ISE_CALIBRATE_SINGLE_REGISTER = 9    # calibration offset */
ISE_CALIBRATE_REFHIGH_REGISTER = 13  # reference high calibration */
ISE_CALIBRATE_REFLOW_REGISTER = 17   # reference low calibration */
ISE_CALIBRATE_READHIGH_REGISTER = 21  # reading high calibration */
ISE_CALIBRATE_READLOW_REGISTER = 25  # reading low calibration */
ISE_SOLUTION_REGISTER = 29          # reference ISE solution */
ISE_BUFFER_REGISTER = 33             # buffer */
ISE_FW_VERSION_REGISTER = 37         # firmware version
ISE_CONFIG_REGISTER = 38             # config register */
ISE_TASK_REGISTER = 39               # task register */

ISE_TEMP_MEASURE_TIME = 750
ISE_MV_MEASURE_TIME = 250

ISE_DUALPOINT_CONFIG_BIT = 0
ISE_TEMP_COMPENSATION_CONFIG_BIT = 1


class uFire_ISE(object):
    mV = 0
    tempC = 0
    tempF = 0
    address = ISE_PROBE

    def __init__(self, address=ISE_PROBE, i2c_bus=3, **kwargs):
        global i2c
        self.address = address
        i2c = smbus.SMBus(i2c_bus)

    def measuremV(self):
        self._send_command(ISE_MEASURE_MV)
        time.sleep(ISE_MV_MEASURE_TIME / 1000.0)
        self.mV = self._read_register(ISE_MV_REGISTER)
        self.mV = round(self.mV * 100.0) / 100.0

        if math.isnan(self.mV):
            self.mV = -1
        if math.isinf(self.mV):
            self.mV = -1

        return self.mV

    def measureTemp(self):
        self._send_command(ISE_MEASURE_TEMP)
        time.sleep(ISE_TEMP_MEASURE_TIME / 1000.0)
        self.tempC = self._read_register(ISE_TEMP_REGISTER)

        if self.tempC == -127.0:
            self.tempF = -127.0
        else:
            self.tempF = ((self.tempC * 9) / 5) + 32

        return self.tempC

    def setTemp(self, temp_C):
        self._write_register(ISE_TEMP_REGISTER, temp_C)
        self.tempC = temp_C
        self.tempF = ((self.tempC * 9) / 5) + 32

    def calibrateSingle(self, solutionmV):
        self._write_register(ISE_SOLUTION_REGISTER, solutionmV)
        self._send_command(ISE_CALIBRATE_SINGLE)
        time.sleep(ISE_MV_MEASURE_TIME / 1000.0)

    def calibrateProbeLow(self, solutionmV):
        self._write_register(ISE_SOLUTION_REGISTER, solutionmV)
        self._send_command(ISE_CALIBRATE_LOW)
        time.sleep(ISE_MV_MEASURE_TIME / 1000.0)

    def calibrateProbeHigh(self, solutionmV):
        self._write_register(ISE_SOLUTION_REGISTER, solutionmV)
        self._send_command(ISE_CALIBRATE_HIGH)
        time.sleep(ISE_MV_MEASURE_TIME / 1000.0)

    def getVersion(self):
        return self._read_byte(ISE_VERSION_REGISTER)

    def getCalibrateOffset(self):
        return self._read_register(ISE_CALIBRATE_SINGLE_REGISTER)

    def getCalibrateHighReference(self):
        return self._read_register(ISE_CALIBRATE_REFHIGH_REGISTER)

    def getCalibrateLowReference(self):
        return self._read_register(ISE_CALIBRATE_REFLOW_REGISTER)

    def getCalibrateHighReading(self):
        return self._read_register(ISE_CALIBRATE_READHIGH_REGISTER)

    def getCalibrateLowReading(self):
        return self._read_register(ISE_CALIBRATE_READLOW_REGISTER)

    def reset(self):
        n = float('nan')
        self._write_register(ISE_CALIBRATE_SINGLE_REGISTER, n)
        self._write_register(ISE_CALIBRATE_REFHIGH_REGISTER, n)
        self._write_register(ISE_CALIBRATE_REFLOW_REGISTER, n)
        self._write_register(ISE_CALIBRATE_READHIGH_REGISTER, n)
        self._write_register(ISE_CALIBRATE_READLOW_REGISTER, n)

    def setDualPointCalibration(self, refLow, refHigh, readLow, readHigh):
        self._write_register(ISE_CALIBRATE_REFLOW_REGISTER, refLow)
        self._write_register(ISE_CALIBRATE_REFHIGH_REGISTER, refHigh)
        self._write_register(ISE_CALIBRATE_READLOW_REGISTER, readLow)
        self._write_register(ISE_CALIBRATE_READHIGH_REGISTER, readHigh)

    def setI2CAddress(self, i2cAddress):
        if i2cAddress >= 1 and i2cAddress <= 127:
            self._write_register(ISE_SOLUTION_REGISTER, float(i2cAddress))
            self._send_command(ISE_I2C)
            self.address = int(i2cAddress)

    def useTemperatureCompensation(self, b):
        retval = self._read_byte(ISE_CONFIG_REGISTER)

        retval = self._bit_set(retval, ISE_TEMP_COMPENSATION_CONFIG_BIT, b)
        self._write_byte(ISE_CONFIG_REGISTER, retval)

    def readEEPROM(self, address):
        self._write_register(ISE_SOLUTION_REGISTER, int(address))
        self._send_command(ISE_MEMORY_READ)
        return self._read_register(ISE_BUFFER_REGISTER)

    def writeEEPROM(self, address, value):
        self._write_register(ISE_SOLUTION_REGISTER, int(address))
        self._write_register(ISE_BUFFER_REGISTER, float(value))
        self._send_command(ISE_MEMORY_WRITE)

    def getFirmware(self):
        return self._read_byte(ISE_FW_VERSION_REGISTER)

    def _bit_set(self, v, index, x):
        mask = 1 << index
        v &= ~mask
        if x:
            v |= mask
        return v

    def _change_register(self, r):
        global i2c
        i2c.write_byte(self.address, r)
        time.sleep(10 / 1000.0)

    def _send_command(self, command):
        global i2c
        i2c.write_byte_data(self.address, ISE_TASK_REGISTER, command)
        time.sleep(10 / 1000.0)

    def _write_register(self, reg, f):
        global i2c
        n = self.round_total_digits(f)
        fd = bytearray(struct.pack("f", n))
        data = [0, 0, 0, 0]
        data[0] = fd[0]
        data[1] = fd[1]
        data[2] = fd[2]
        data[3] = fd[3]
        self._change_register(reg)
        i2c.write_i2c_block_data(self.address, reg, data)
        time.sleep(10 / 1000.0)

    def _read_register(self, reg):
        global i2c
        data = [0, 0, 0, 0]
        self._change_register(reg)
        data[0] = i2c.read_byte(self.address)
        data[1] = i2c.read_byte(self.address)
        data[2] = i2c.read_byte(self.address)
        data[3] = i2c.read_byte(self.address)
        ba = bytearray(data)
        f = struct.unpack('f', ba)[0]
        return self.round_total_digits(f)

    def _write_byte(self, reg, val):
        global i2c
        i2c.write_byte_data(self.address, reg, val)
        time.sleep(10 / 1000.0)

    def _read_byte(self, reg):
        global i2c
        self._change_register(reg)
        time.sleep(10 / 1000.0)
        return i2c.read_byte(self.address)

    def magnitude(self, x):
        if math.isnan(x):
            return 0
        return 0 if x == 0 else int(math.floor(math.log10(abs(x)))) + 1

    def round_total_digits(self, x, digits=7):
        return round(x, digits - self.magnitude(x))
